### PR TITLE
Add integration test for External Node

### PR DESCRIPTION
**Pull request by archanapholla**

### CODE REVIEW

File: test/integration/externalnode_test.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+	"fmt"
+	"math/rand"
+	"path"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	"github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
```
Comment (Anandkumar26) [inline test/integration/externalnode_test.go:28]: Please fix the imports.

`corev1 "k8s.io/api/core/v1"`
`metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"`

File: test/integration/externalnode_test.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+	v1 "k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	antreatypes "antrea.io/antrea/pkg/apis/crd/v1alpha1"
```
Comment (Anandkumar26) [inline test/integration/externalnode_test.go:34]: Prefer to use it as 

`antreav1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"`
`antreav1alpha2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"`

File: test/utils/cloud_provider_aws.go [hunk 3]
```diff
@@ -130,6 +130,10 @@ func (p *awsVPC) GetVMNames() []string {
 	return vmNames
 }
 
+func (p *awsVPC) GetPrimaryNICs() []string {
```
Comment (Anandkumar26) [inline test/utils/cloud_provider_aws.go:133]: Is this required? Can't we re-use GetNICs() function

Comment (archanapholla) [reply test/utils/cloud_provider_aws.go:133]: No, Required for Azure case, comments have been added for pointing the difference between the 2 functions. GetPrimaryNICs() was required to be added to return the complete ID of NIC, GetNICs() returns just the name.

File: go.mod [hunk 4]
```diff
@@ -101,6 +101,8 @@ require (
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
+require github.com/seancfoley/ipaddress-go v1.2.1
```
Comment (reachjainrahul) [inline go.mod:104]: give this to line 113 and remove indirect from there

File: test/integration/externalnode_test.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+	}
+
+	deleteCloudEntitySelector := func() {
+		logf.Log.Info("Delete", "selector", selector.Name)
+		err := k8sClient.Delete(context.TODO(), selector)
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	validateResult := func(expectedResult string, kubectlCmd string, agented bool, en *antreav1alpha1.ExternalNode,
+		ee *antreav1alpha2.ExternalEntity) {
```
Comment (reachjainrahul) [inline test/integration/externalnode_test.go:125]: nit: align with function arguments?

File: go.mod (hunk 4 above)
Comment (archanapholla) [reply go.mod:104]: not sure why this happened, i reverted it now.

File: test/integration/externalnode_test.go (hunk 5 above)
Comment (archanapholla) [reply test/integration/externalnode_test.go:125]: inteliJ is not allowing.

File: test/integration/externalnode_test.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+		}
+	}
+
+	BeforeEach(func() {
+		if preserveSetupOnFail {
+			preserveSetup = true
+		}
+		namespace = &corev1.Namespace{}
+		namespace.Name = namespaceName + "-" + fmt.Sprintf("%x", rand.Int())
+		getEnCmd = fmt.Sprintf("get externalnode -n %s -o=jsonpath='{range.items[*]}{.metadata.name}{\"\\n\"}{end}'", namespace.Name)
```
Comment (reachjainrahul) [inline test/integration/externalnode_test.go:253]: use getEnCmd directly at line 233. avoid variable use

File: test/integration/externalnode_test.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+		getEeCmd = fmt.Sprintf("get externalentity -n %s -o=jsonpath='{range.items[*]}{.metadata.name}{\"\\n\"}{end}'", namespace.Name)
+		vmIdToMatch = cloudVPC.GetVMIDs()[0]
+		testVm = cloudVPC.GetVMs()[0]
+		nic = cloudVPC.GetPrimaryNICs()[0]
+		publicIp = cloudVPC.GetVMIPs()[0]
+		privateIp = cloudVPC.GetVMPrivateIPs()[0]
```
Comment (reachjainrahul) [inline test/integration/externalnode_test.go:259]: nit: new line.. Group relevant initialization together. separate by new line

File: test/integration/externalnode_test.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+			}
+			logf.Log.Error(ret, "External Node not removed", "fetch key", key)
+			return false, ret
+		})
+		return err
+	}
+
+	tester := func(expectedResult string, selectorExists bool, agented bool, isDelete bool) {
+		var kubectlCmd string
+		en := &antreav1alpha1.ExternalNode{}
```
Comment (reachjainrahul) [inline test/integration/externalnode_test.go:215]: this can go at line 227.

File: test/integration/externalnode_test.go [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+			}
+			return true, nil
+		})
+		return err
+	}
+
+	checkEeCreated := func(key client.ObjectKey, ee *antreav1alpha2.ExternalEntity) error {
+		logf.Log.Info("Check for External Entity creation")
+		err := wait.Poll(time.Second*5, time.Second*60, func() (bool, error) {
+			ret := k8sClient.Get(context.TODO(), key, ee)
```
Comment (reachjainrahul) [inline test/integration/externalnode_test.go:190]: if the object is not present, does it return error. if so, we will be logging error on every call. Verify and avoid 192 logging. have a final check on 197 to say failure

File: test/integration/externalnode_test.go [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+	deleteAccount := func() {
+		logf.Log.Info("Delete", "account", account.Name)
+		err := k8sClient.Delete(context.TODO(), account)
+		Expect(err).ToNot(HaveOccurred())
+		err = k8sClient.Delete(context.TODO(), secret)
+		Expect(err).ToNot(HaveOccurred())
+	}
+
+	applyCloudEntitySelector := func(selectorExists bool, agented bool) {
```
Comment (reachjainrahul) [inline test/integration/externalnode_test.go:93]: Also add the test coverage for other match like VPC ID.

File: test/integration/integration_suite_test.go [hunk 11]
```diff
@@ -32,7 +32,8 @@ import (
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
-	antreatypes "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	antreatypes "antrea.io/antrea/pkg/apis/crd/v1alpha1"
```
Comment (Anandkumar26) [inline test/integration/integration_suite_test.go:35]: Can you rename these imports as well?

File: test/integration/externalnode_test.go [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+		deleteAccount()
+		deleteNS()
+	})
+
+	table.DescribeTable("Testing agented and agentless configuration through CES",
+		func() {
+			expectedResult := fmt.Sprintf("virtualmachine-%s", testVm)
+
+			By("Configure entity selector with agented field set to true, check ExternalNode creation")
+			tester(expectedResult, false, true, false)
```
Comment (Anandkumar26) [inline test/integration/externalnode_test.go:347]: We should be checking for the ExternalNode deletion, before running this test.
i.e. order should be something like this

1. Configure EntitySelector with Agented=true for VM1, verify En is created.
2. Configure EntitySelector with Agented=false for VM1, verify Ee is created, verify En is deleted.
3. Configure EntitySelector with Agented=true for VM1, verify En is created, verify Ee from step 2 is deleted.
4. Delete EntitySelector and verify En Deletion.

It seems that currently we are missing additional verification in 2nd and 3rd step.

File: test/integration/externalnode_test.go (hunk 6 above)
Comment (archanapholla) [reply test/integration/externalnode_test.go]: Ok

File: test/integration/externalnode_test.go (hunk 7 above)
Comment (archanapholla) [reply test/integration/externalnode_test.go:295]: done

File: test/integration/externalnode_test.go (hunk 9 above)
Comment (archanapholla) [reply test/integration/externalnode_test.go:187]: Returning error from function and logging on the caller on error. Also, made sure that failure is not returned on poll failure within the timeout

File: test/integration/externalnode_test.go (hunk 10 above)
Comment (archanapholla) [reply test/integration/externalnode_test.go]: We already have entity selector test covering VPC ID testing match selector, I thought the focus of this test is to test external node lifecycle, hence i didn't want to add different match criteria in CES.
If you think we need to have VPC ID in this test too, i will take up soon after the ongoing task.

File: test/integration/integration_suite_test.go (hunk 11 above)
Comment (archanapholla) [reply test/integration/integration_suite_test.go]: done

File: test/integration/externalnode_test.go [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"k8s.io/apimachinery/pkg/util/wait"
```
[the remark this answers is not in the document]
Comment (archanapholla) [reply test/integration/externalnode_test.go]: got introduced by itself, corrected it.

File: test/integration/externalnode_test.go (hunk 12 above)
Comment (archanapholla) [reply test/integration/externalnode_test.go:384]: Done

**Review by reachjainrahul — APPROVED** [review]
/LGTM
@Anandkumar26 Merge the changes if you are comfortable too.

File: test/integration/externalnode_test.go (hunk 10 above)
Comment (reachjainrahul) [reply test/integration/externalnode_test.go]: Yes, we need test for all combination, VPC ID, VM ID, VMName and VPC Name. As the External Node Cache indexer is different from External Entity, we cannot rely on those tests. you can take it up next

**Review by Anandkumar26 — APPROVED** [review]
/LGTM

nephe-test-e2e-aks - failed, which is not related to this PR.
Applied NetworkPolicy to VirtualMachine by kind label selector using group
failed to poll policies [test-cloud-setup-anp test-cloud-anp] for resources [ubuntu-host-ci-vmlinux-0-14148]: timed out waiting for the condition

**Pull request merged**